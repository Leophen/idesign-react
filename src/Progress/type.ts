export interface ProgressProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 进度条百分比
   * @default 0
   */
  percentage?: number;
  /**
   * 进度条类型
   * @default bar
   */
  type?: 'bar' | 'circle';
  /**
   * 是否显示进度提示文本
   * @default true
   */
  labelTxt?: boolean;
  /**
   * 提示是否在进度条内部
   * @default false
   */
  innerLabel?: boolean;
  /**
   * 自定义进度提示
   */
  label?: React.ReactNode;
  /**
   * 进度条颜色
   */
  color?: string;
  /**
   * 进度条底色
   */
  backColor?: string;
  /**
   * 进度条长度/环形直径
   * @default 300
   */
  width?: React.CSSProperties["width"];
  /**
   * 进度条粗细度
   * @default 6
   */
  strokeWidth?: number;
  /**
   * 触发进度加载动画
   * @default false
   */
  indeterminate?: boolean;
  /**
   * 进度加载动画速度
   * @default 3
   */
  duration?: number;
}
