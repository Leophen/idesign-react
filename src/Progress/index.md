---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 15
---

# Progress 进度条

用于显示进度的进度条组件。

## 基本用法

通过 `percentage` 属性控制进度。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div>
      <Progress percentage={60} />
    </div>
  );
};

export default App;
```

## 隐藏进度提示

通过 `labelTxt` 属性控制是否显示进度提示。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div>
      <Progress percentage={60} labelTxt={false} />
    </div>
  );
};

export default App;
```

## 进度提示在内

条形进度条通过 `innerLabel` 属性控制是否显示进度提示，注意使用该属性时进度条最小高度会设为 `20`。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div>
      <Progress percentage={60} innerLabel />
    </div>
  );
};

export default App;
```

## 自定义进度提示

通过 `label` 属性设置自定义进度提示。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div className="i-design-demo-column">
      <Progress percentage={30} label="😄" />
      <Progress percentage={60} label={<div>🌛</div>} />
    </div>
  );
};

export default App;
```

## 自定义进度条颜色

通过 `color` 属性设置自定义进度条颜色，`backColor` 属性控制进度条底色。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div className="i-design-demo-column">
      <Progress percentage={30} color="#FFCF2E" />
      <Progress percentage={60} color="linear-gradient(90deg, rgb(0, 255, 255), rgb(255, 0, 255))" backColor="#FFCF2E" />
    </div>
  );
};

export default App;
```

## 自定义宽度或粗细

通过 `width` 来控制条形进度条长度/环形进度条直径，通过 `strokeWidth` 属性控制进度条粗细度。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Progress percentage={30} width={300} />
      <Progress percentage={60} strokeWidth={20} />
    </div>
  );
};

export default App;
```

## 环形进度条

通过 `type` 属性设置进度条类型，为 `circle` 时显示环形进度条。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Progress type="circle" percentage={0} />
      <Progress type="circle" percentage={30} width={100} />
      <Progress type="circle" percentage={60} width={100} strokeWidth={20} />
      <Progress type="circle" percentage={90} width={100} label={<div>🌛</div>} />
    </div>
  );
};

export default App;
```

## 进度加载动画

通过 `indeterminate` 属性设置不确定的进度加载循环动画，可用 `duration` 控制循环速度，默认为 `3`。

```tsx
import React from 'react';
import { Progress } from 'idesign-react';

const App = () => {
  return (
    <>
      <div className="idesign-demo-block-column">
        <Progress percentage={30} indeterminate />
        <Progress percentage={30} indeterminate duration={1} />
      </div>
      <div className="idesign-demo-block-row" style={{ marginTop: 20 }}>
        <Progress type="circle" width={100} indeterminate />
        <Progress type="circle" width={100} percentage={30} indeterminate />
        <Progress type="circle" width={100} duration={1} indeterminate />
      </div>
    </>
  );
};

export default App;
```

<API />
