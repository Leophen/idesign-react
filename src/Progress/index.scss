.i-progress {
  display: flex;
  align-items: center;
}

@keyframes indeterminate_bar {
  0% {
    left: -100%;
  }
  to {
    left: 100%;
  }
}

.i-progress-bar {
  position: relative;
  overflow: hidden;
  background: var(--i-bg-back);
  border-radius: 100px;
  .i-progress__info {
    position: absolute;
    top: 50%;
    right: 6px;
    transform: translateY(-50%);
  }
  &.i-progress-bar__has-label {
    min-height: 20px;
  }
}

.i-progress-circle__bg {
  stroke: var(--i-bg-back);
}

.i-progress-bar__inner {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  background: var(--i-primary);
  border-radius: 100px;
}

.i-progress-circle__inner {
  stroke: var(--i-primary);
}

.i-progress__info {
  margin-left: 8px;
  color: var(--i-font-2);
  font-size: 14px;
}

@keyframes indeterminate_circle {
  to {
    stroke-dashoffset: 0;
  }
}

.i-progress-circle {
  position: relative;
  svg {
    transform: rotate(-90deg);
  }
  .i-progress__info {
    position: absolute;
    top: 50%;
    left: 50%;
    margin: 0;
    transform: translate(-50%, -50%);
  }
}
