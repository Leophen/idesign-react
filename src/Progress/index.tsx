import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { ProgressProps } from './type';

const Progress: React.FC<ProgressProps> = (props) => {
  const {
    className,
    style,
    percentage = 0,
    type = 'bar',
    labelTxt = true,
    innerLabel = false,
    label,
    color,
    backColor,
    width = 200,
    strokeWidth = 6,
    indeterminate = false,
    duration = 3,
    ...restProps
  } = props;

  const animateBar = `indeterminate_bar ${duration}s infinite ease-in-out`

  const progressBar = (
    <>
      <div
        className={classNames(
          'i-progress-bar',
          innerLabel && 'i-progress-bar__has-label',
        )}
        style={{
          width,
          height: strokeWidth,
          background: backColor
        }}
      >
        <div
          className="i-progress-bar__inner"
          style={{
            width: `${percentage}%`,
            background: color,
            animation: indeterminate ? animateBar : 'unset'
          }}
        />
        {innerLabel && (labelTxt || label) && (
          <div className="i-progress__info">
            {!label ? `${percentage}%` : label}
          </div>
        )}
      </div>
      {!innerLabel && (labelTxt || label) && (
        <div className="i-progress__info">
          {!label ? `${percentage}%` : label}
        </div>
      )}
    </>
  )

  const getStyleNum = (val: string | number) => {
    return parseInt(String(width))
  }

  const circle = {
    d: getStyleNum(width),
    r: (getStyleNum(width) / 2) - strokeWidth / 2,
    l: getStyleNum(width) * Math.PI
  }

  const animateCircle = `indeterminate_circle ${duration}s infinite linear`

  const progressCircle = (
    <div
      className={classNames(
        'i-progress-circle',
        innerLabel && 'i-progress-circle__has-label',
      )}
      style={{
        width,
        height: width
      }}
    >
      <svg
        width={circle.d}
        height={circle.d}
      >
        <circle
          cx={circle.d / 2}
          cy={circle.d / 2}
          r={circle.r}
          fill="none"
          className="i-progress-circle__bg"
          style={{
            stroke: backColor,
            strokeWidth,
          }}
        />
        <circle
          cx={circle.d / 2}
          cy={circle.d / 2}
          r={circle.r}
          fill="none"
          className="i-progress-circle__inner"
          style={{
            stroke: color,
            strokeWidth,
            strokeLinecap: 'round',
            strokeDasharray: circle.l,
            strokeDashoffset: !indeterminate ? circle.l - circle.l * percentage / 100 : circle.l * 2,
            animation: indeterminate ? animateCircle : 'unset'
          }}
        />
      </svg>
      {(labelTxt || label) && (
        <div className="i-progress__info">
          {!label ? `${percentage}%` : label}
        </div>
      )}
    </div>
  )

  return (
    <div
      className={classNames(
        'i-progress',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      {type !== 'bar' ? progressCircle : progressBar}
    </div>
  );
};

Progress.displayName = 'Progress';

export default Progress;
