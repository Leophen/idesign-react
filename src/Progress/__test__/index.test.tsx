import { render } from '@testing-library/react';
import Progress from '../index';

describe('Progress ç»„ä»¶æµ‹è¯•', () => {
  it('create', () => {
    const { asFragment } = render(<Progress percentage={60} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('label', () => {
    const { queryByText } = render(<Progress label="ğŸ˜„" />);
    expect(queryByText('ğŸ˜„')).toBeInTheDocument();
  });

  it('hide label', () => {
    const { container } = render(<Progress labelTxt={false} />);
    expect(container.querySelector('.i-progress__info')).not.toBeInTheDocument();
  });

  it('innerLabel', () => {
    const { container } = render(<Progress innerLabel />);
    expect(container.firstChild.firstChild).toHaveClass('i-progress-bar__has-label');
  });

  it('color backColor', () => {
    const { container } = render(<Progress color="blue" backColor="red" />);
    expect(container.querySelector('.i-progress-bar')).toHaveStyle('background: red');
    expect(container.querySelector('.i-progress-bar__inner')).toHaveStyle('background: blue');
  });

  it('width', () => {
    const { container } = render(<Progress width={300} />);
    expect(container.firstChild.firstChild).toHaveStyle('width: 300px');
  });

  it('strokeWidth', () => {
    const { container } = render(<Progress strokeWidth={30} />);
    expect(container.firstChild.firstChild).toHaveStyle('height: 30px');
  });

  it('circle', () => {
    const { asFragment } = render(<Progress type="circle" percentage={60} width={100} strokeWidth={20} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

