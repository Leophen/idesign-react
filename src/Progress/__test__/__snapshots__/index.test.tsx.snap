// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress 组件测试 circle 1`] = `
<DocumentFragment>
  <div
    class="i-progress"
  >
    <div
      class="i-progress-circle"
      style="width: 100px; height: 100px;"
    >
      <svg
        height="100"
        width="100"
      >
        <circle
          class="i-progress-circle__bg"
          cx="50"
          cy="50"
          fill="none"
          r="40"
          style="stroke-width: 20;"
        />
        <circle
          class="i-progress-circle__inner"
          cx="50"
          cy="50"
          fill="none"
          r="40"
          style="stroke-width: 20; stroke-linecap: round; stroke-dasharray: 314.1592653589793; stroke-dashoffset: 125.66370614359175; animation: unset;"
        />
      </svg>
      <div
        class="i-progress__info"
      >
        60%
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Progress 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-progress"
  >
    <div
      class="i-progress-bar"
      style="width: 200px; height: 6px;"
    >
      <div
        class="i-progress-bar__inner"
        style="width: 60%; animation: unset;"
      />
    </div>
    <div
      class="i-progress__info"
    >
      60%
    </div>
  </div>
</DocumentFragment>
`;
